src/osd/scrubber/scrub_backend.h

```diff
@@ -0,0 +1,419 @@
+  ScrubBackend(PgScrubber& scrubber,
+               PGBackend& backend,
+               PG& pg,
+               pg_shard_t i_am,
+               bool repair,
+               scrub_level_t shallow_or_deep,
+               const std::set<pg_shard_t>& acting);
+
+  // Replica constructor: no primary map
+  ScrubBackend(PgScrubber& scrubber,
+               PGBackend& backend,
+               PG& pg,
+               pg_shard_t i_am,
+               bool repair,
+               scrub_level_t shallow_or_deep);
+
+  friend class PgScrubber;
+
+  /**
+   * reset the per-chunk data structure (ScrubberBeChunk).
+   * Create an empty scrub-map for this shard, and place it
+   * in the appropriate entry in 'received_maps'.
+   *
+   * @returns a pointer to the newly created ScrubMap.
+   */
+  ScrubMap* new_chunk() final;
+
+  /**
+   * sets Backend's m_repair flag (setting m_mode_desc to a corresponding
+   * string)
+   */
+  void update_repair_status(bool should_repair) final;
+
+  // note: used by both Primary & replicas
+  static ScrubMap clean_meta_map(ScrubMap& cleaned, bool max_reached);
+
+  void replica_clean_meta(ScrubMap& smap,
+                          bool max_reached,
+                          const hobject_t& start) final;
+
+  /**
+   * decode the arriving MOSDRepScrubMap message, placing the replica's
+   * scrub-map into received_maps[from].
+   *
+   * @param from replica
+   * @param pool the pool the replica is in
+   */
+  void decode_received_map(pg_shard_t from, const MOSDRepScrubMap& msg) final;
+
+  void scrub_compare_maps(bool max_reached) final;
+
+  int get_num_digest_updates_pending() const final
+  {
+    return num_digest_updates_pending;
+  }
+
+  void scan_snaps(ScrubMap& smap) final;
+
+  int scrub_process_inconsistent() final;
+
+  void repair_oinfo_oid(ScrubMap& smap) final;
+
+  std::ostream& logger_prefix(std::ostream* _dout, const ScrubBackend* t);
+
+ private:
+  // set/constructed at the ctor():
+  PgScrubber& m_scrubber;
+  PGBackend& m_pgbe;
+  PG& m_pg;
+  const pg_shard_t m_pg_whoami;
+  bool m_repair;
+  const scrub_level_t m_depth;
+  const spg_t m_pg_id;
+  std::vector<pg_shard_t> m_acting_but_me;  // primary only
+  bool m_is_replicated{true};
+  std::string_view m_mode_desc;
+  std::string m_formatted_id;
+
+  // shorthands:
+  ConfigProxy& m_conf;
+  LogChannelRef clog;
+
+  int num_digest_updates_pending{0};
+
+ public:  
+  // as used by PgScrubber::final_cstat_update(). consider relocating.
+  // actually - only filled in by the PG backend, and used by the scrubber.
+  // We are not handling it. So consider getting it from the Scrubber, or
+  // creating it by the PG-BE
+  omap_stat_t m_omap_stats = (const struct omap_stat_t){0};
+
+ private:
+
+  using AuthAndObjErrors = std::tuple<std::list<pg_shard_t>,  ///< the auth-list
+                                      std::set<pg_shard_t>    ///< object_errors
+                                      >;
+
+  std::optional<ScrubBeChunk> this_chunk;
```

Comment:
new_chunk() has a side effect of clearing this_chunk.  It's not clear to me where m_missing or m_inconsistent get cleared.  I don't really understand why PGScrubber doesn't just create a new instance of this entire class for each chunk -- there doesn't seem to be any state intended to survive between chunks.

Reply:
I forgot to rm the old m_missing (BTW - clearing was not its problem... it was just cleared, with no other usage). Thanks for spotting that.

The whole scrub backend is created anew each scrub.